Sources/AblyChat/ChatAPI.swift

```diff
@@ -0,0 +1,232 @@
+                            continuation.resume(returning: decodedResponse)
+                        } catch {
+                            continuation.resume(throwing: error)
+                        }
+                    })
+                } catch {
+                    continuation.resume(throwing: error)
+                }
+            }
+    }
+    
+    private func makeAuthorizedPaginatedRequest<RES: Codable & Sendable>(
+        _ url: String,
+        params: [String: String]? = nil,
+        body: [String: Any]? = nil
+    ) async throws -> any PaginatedResult<RES> {
+        if #available(macOS 13.0.0, *) {
+            return try await withCheckedThrowingContinuation { continuation in
+                do {
+                    try rest.request("GET", path: url, params: params, body: nil, headers: ["protocol": String(apiProtocolVersion)], callback: { paginatedResponse, error  in
+                        
+                        ARTHTTPPaginatedCallbackWrapper<RES>(callback: (paginatedResponse, error)).handleResponse(continuation: continuation)
+                    })
+                } catch {
+                    continuation.resume(throwing: error)
+                }
+            }
+            
+        } else {
+            // Fallback on earlier versions
+            throw ARTErrorInfo.create(withCode: 501, status: 0, message: "Unsupported macOS version")
+        }
+    }
+}
+
+@available(macOS 13.0.0, *)
+struct ARTHTTPPaginatedCallbackWrapper<RES: Codable & Sendable> {
+    let callback: (ARTHTTPPaginatedResponse?, ARTErrorInfo?)
+    
+    init(callback: (ARTHTTPPaginatedResponse?, ARTErrorInfo?)) {
+        self.callback = callback
+    }
+    
+    func handleResponse(continuation: CheckedContinuation<any PaginatedResult<RES>, any Error>) {
+        if let error = callback.1 {
+            continuation.resume(throwing: ARTErrorInfo.create(withCode: error.code, status: error.statusCode, message: error.message))
+            return
+        }
+        
+        guard let paginatedResponse = callback.0, paginatedResponse.statusCode == 200 else {
+            continuation.resume(throwing: ARTErrorInfo.createUnknownError())
+            return
+        }
+        
+        do {
+            let jsonData = try JSONSerialization.data(withJSONObject: paginatedResponse.items)
+            let decodedData = try JSONDecoder().decode([RES].self, from: jsonData)
+            let result = paginatedResponse.toPaginatedResult(items: decodedData)
+            continuation.resume(returning: result)
+        } catch {
+            continuation.resume(throwing: error)
+        }
+    }
+}
+
+extension ARTHTTPPaginatedResponse {
+    func toPaginatedResult<T: Codable & Sendable>(items: [T]) -> PaginatedResultWrapper<T>{
+        
+        return PaginatedResultWrapper(paginatedResponse: self, items: items)
+    }
+}
+
+public final class PaginatedResultWrapper<T: Codable & Sendable>: PaginatedResult {
```

Comment:
This can go in a separate file, I think.

Reply:
done